src/v/cloud_storage/tests/segment_meta_cstore_test.cc

```diff
@@ -1342,3 +1344,551 @@ BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_append_retrieve_edge_case) {
+            return;
+        }
+        auto b_it = cstore.find(first);
+        auto e_it = cstore.end();
+        BOOST_REQUIRE(cond1(b_it == cstore.begin()));
+        BOOST_REQUIRE(cond1(*cstore.prev(e_it) == cstore.last_segment()));
+    };
+
+    // Single batch invariants
+    auto check_local_invariants = [&](
+                                    const segment_meta& meta,
+                                    bool is_replacement) {
+        vlog(test.info, "Checking {}", meta);
+        auto o = meta.base_offset;
+        // check that it can be found by its base offset
+        {
+            auto it = cstore.find(o);
+            BOOST_REQUIRE(cond(it != cstore.end(), meta, is_replacement));
+            BOOST_REQUIRE(cond(*it == meta, meta, is_replacement));
+        }
+
+        // check that it can be found using upper_bound
+        {
+            auto it = cstore.upper_bound(o);
+            if (it != cstore.begin()) {
+                auto it2 = cstore.prev(it);
+                auto m2 = *it2;
+                BOOST_REQUIRE(cond(walk.is_valid(m2), m2, is_replacement));
+                BOOST_REQUIRE(cond(m2.base_offset == o, meta, is_replacement));
+            } else {
+                BOOST_REQUIRE(cond(cstore.size() == 1, meta, is_replacement));
+            }
+        }
+    };
+    for (int i = 0; i < n_iter; i++) {
+        auto [t_off, exp_so] = walk.maybe_truncate();
+        if (t_off != model::offset{}) {
+            vlog(
+              test.info, "Cstore truncate {}, expected SO: {}", t_off, exp_so);
+            cstore.prefix_truncate(t_off);
+            check_global_invariants(exp_so);
+        }
+        auto [meta, op_type] = walk.next_segment();
+
+        try {
+            vlog(test.info, "Cstore insert {}", meta);
+            auto tx = cstore.append(meta);
+            std::ignore = std::move(tx);
+        } catch (const std::invalid_argument& e) {
+            vlog(
+              test.info,
+              "Cstore insert {} failed with {}, retrying with flush",
+              meta,
+              e.what());
+            cstore.flush_write_buffer();
+            auto tx = cstore.append(meta);
+            std::ignore = std::move(tx);
+        }
+
+        switch (op_type) {
+        case segment_meta_random_walk::op_type::append:
+            if (model::test::with_probability(0.1)) {
+                check_last_element_invariants(meta);
+            }
+            break;
+        case segment_meta_random_walk::op_type::prepend:
+            check_global_invariants(walk.get_start_offset().value());
+            break;
+        case segment_meta_random_walk::op_type::replace:
+            if (model::test::with_probability(0.1)) {
+                check_local_invariants(meta, true);
+                for (const auto& [key, m] : walk.get_segments()) {
+                    std::ignore = key;
+                    check_local_invariants(m, true);
+                }
+            }
+            break;
+        };
```

Comment:
Is it prohibitively expensive to run local and global checks on every iteration?